rs/messaging/src/routing/stream_handler/tests.rs

```
@@ -2384,10 +2388,11 @@ fn induct_stream_slices_partial_success() {
                 // ...a request to a missing canister @46 (on this subnet according to the
                 // routing table); this is expected to trigger a reject response...
                 Request(*REMOTE_CANISTER, *OTHER_LOCAL_CANISTER),
-                // ..a request not from `REMOTE_SUBNET` @47...
+                // ..a request not from `REMOTE_SUBNET` @47; this is also expected to trigger
+                // a reject respose...
                 Request(*LOCAL_CANISTER, *LOCAL_CANISTER),
                 // ...a request from a missing canister @48 (not anywhere according to the routing
-                // table); this is expected to be accepted but dropped...
+                // table); this is expected to be a reject response...
                 Request(*UNKNOWN_CANISTER, *LOCAL_CANISTER),
```

In your other PR, I seem to remember that we still treat a request addressed **to** a canister not assigned to any subnet as a critical error. I wonder what the correct behavior should be.

If a new subnet (with newly assigned canister ranges) does pop into existence and "this subnet" has not yet learned of its existence, could some canister on said subnet have already migrated? I suppose so.

Could that canister be migrating to "this subnet"? (In the case of the other PR.) I guess that depends exactly on the implementation details of manual canister migration (i.e. when exactly we update the routing table). But I'm not actually sure. And the behavior may technically change with a "minor" change to said implementation. @schneiderstefan , any opinion?

---

Technically I guess that's true. After changing the routing table, we do wait for source and target subnet to update their registry version. But of course it is possible that during that time, a third subnet learns of the registry change first, and sends a message. The target subnet might then still be so far behind as to not even know about the new subnet. 

But this is a very outlandish scenario. Somebody must have installed a canister on the new subnet, triggered a canister migration, all the checks of cycle balances, routing tables etc must have happened, and yet the target subnet still doesn't know about the new subnet. Since the only bad thing that happens is that we get an alert we shouldn't have, my instinct tells me we should not overengineer anything around this.

---

It's not only an alert. The request is silently dropped, so we leak a callback and a canister becomes (the wrong kind of) unstoppable. Also, it could be that the target subnet was stalled for hours, while the whole sequence of events you described was unfolding. So not actually impossible.

I'm starting to think that it may be safer to treat all non-local canisters the same (even though I'd prefer not to).

---

If we go with treating both "canister hosted by another subnet" and "canister not hosted by any known subnet" the same, yes.

But more fundamentally, this is about being consistent. Or making an explicit, reasoned decision for treating the "canister not hosted by any known subnet" case differently when it's a message routed to said canister vs a message from said canister.

---

Incoming requests __from__ the "wrong" subnet we do indeed treat in the same way (this PR). It's only incoming requests __to__ the wrong subnet that we don't treat uniformly 
(#6430).

I'm not sure about your claim about responses. Actually, looking at `induct_stream_slices_partial_success()` I don't see us actually testing responses from canisters that are supposed to be on another subnet; we only test responses from canisters not supposed to be hosted by any subnet. Could you please add the missing case?

---

PTAL

---

Hmm... Actually I'm being stupid. We didn't test for any kind of response coming from the wrong subnet (whether a different subnet or no known subnet). We only tested for a canister that was supposed to be hosted by this subnet but did not exist.

Could you please add both cases to the test?

---

I said "we only test responses from canisters not supposed to be hosted by any subnet", but not for "responses from canisters that are supposed to be on another subnet".

Turns out we had neither, we only tested for a response resulting in `CanisterNotFound` (i.e. canister assigned to "this" subnet, but not present on the subnet).

So you already added a "response from a canister that are supposed to be on another subnet". Please also add a "response from a canister not supposed to be hosted by any subnet".

---

Hm, I thought I added a response not supposed to be anywhere, as the comment about request @50 says. We have `LOCAL_SUBNET`, `REMOTE_SUBNET` and `CANISTER_MIGRATION_SUBNET`; should I add and setup `OTHER_REMOTE_SUBNET` for this testcase?

---

Should I add `ANOTHER_REMOTE_SUBNET`? Then there's a clash in naming cause we have `ANOTHER_REMOTE_CANISTER` that I should alos rename throughout the tests